File: JavaScript/Array/reverse-an-array.js
/*
// Two-Pointer Approach to reverse an array in place:
// 1. start = 0, end = array.length - 1
// 2. Swap array[start] and array[end]
// 3. Increment start, decrement end
// 4. Repeat until start >= end
*/

const numbers = [1,2,3,4,5,6];

function reverseArray(numbers){
    start = 0;
    end = numbers.length - 1;
  while(start < end){
    [numbers[start],numbers[end]] = [numbers[end],numbers[start]];
    start += 1;
    end -= 1;
  }
}

// console.log( `After Reverse: ${numbers.join(" ")}`);
reverseArray(numbers)
//  console.log( `Before Reverse: ${numbers.join(" ")}`);

//We use join
function printArray(numbers){
    console.log(numbers.join(" "));
}

printArray(numbers); // Output: 1 2 3 4 5


//In Js we could also use bilt-in method array.reverse()
console.log(numbers.reverse()); // reverses array in place

File: C++/Array/reverse-an-array.cpp
/*
Problem Statement
You are given an array of numbers. Your task is to reverse the elements of the array.
Original array: [1, 2, 3, 4, 5]
Reversed array: [5, 4, 3, 2, 1]

I am using here two-pointer approach
// Two-Pointer Approach to reverse an array:
// 1. Initialize two pointers: start = 0, end = n-1
// 2. Swap elements at start and end
// 3. Move start forward and end backward
// 4. Repeat until start >= end
// This reverses the array in place with O(n) time and O(1) space
*/

#include<iostream>
using namespace std;

void reversearray(int numbers[],int n){
    int start = 0;
    int end = n - 1;

    while (start < end){
        swap(numbers[start],numbers[end]);
        start +=1;
        end -=1;
    }
}
//show the array
void showarray(int numbers[],int n){
    for(int i=0; i<n;i++){
        cout<<numbers[i]<<" ";
    }
}

int main(){
 int numbers[] = {1,2,3,4,5};
 int n = sizeof(numbers)/sizeof(numbers[0]);

 cout<<"Before Reverse"<<endl;
 showarray(numbers,n);
 cout<<endl;
 cout<<"After Reverse"<<endl;
 reversearray(numbers,n);
 showarray(numbers,n);

    return 0;
}
